File: src/importer.cpp
#include "importer.hpp"

#include "../lib/namedfstream.hpp"

//PRIVATE METHODS

std::string
Importer::merge_filestreams(std::vector<CadumpLibs::NamedFstream *> filestreams,
                            std::string separator) {
  std::string return_string = "";
  for (int i = 0; i < filestreams.size(); i++) {
    // Get lenght of the file
    filestreams[i]->seekg(0, filestreams[i]->end);
    int lenght = filestreams[i]->tellg();
    filestreams[i]->seekg(0, filestreams[i]->beg);

    char read_block[lenght];
    filestreams[i]->read(read_block, lenght);

    if ( i < filestreams.size() - 1)
      return_string += read_block + separator;
    else
      return_string += read_block;
  }

  return return_string;
}

//PUBLIC METHODS

Importer::Importer(CadumpLibs::NamedFstream *entry_file) {
  // TODO: add this when async
  // std::hash<std::thread::id>{}(std::this_thread::get_id()) generate truly
  // unique file name based on pseudorandom number
  master_file_.path = "./temp/temp" + std::to_string(rand());

  master_file_.rw = new CadumpLibs::NamedFstream(
      master_file_.path.c_str(), std::fstream::in | std::fstream::out);

  this->entry_file_ = entry_file;
}

CadumpLibs::NamedFstream *Importer::import() {
  while (!entry_file_->eof()) {
    char line_char[500];
    entry_file_->getline(line_char, 500);
    std::string line(line_char);

    if (line.find("import") != std::string::npos) {
      std::string import_command = get_whole_import_command(line);

      ImportList import_list;

      std::vector<CadumpLibs::NamedFstream *> child_import_list = {};

      if (!import_command.empty()) {
        import_list = parse_imports(import_command);

        for (uint i = 0; i < import_list.size(); i++) {
          import_list.at(i) = satisfy_import_scope(import_list.at(i));

          child_import_list.push_back(spawn_if_doesnt_exist(import_list.at(i)));
        }

        write_to_master_file(merge_filestreams(child_import_list, "\n"));
      }
    }

    master_file_.rw->close();
    master_file_.r = new CadumpLibs::NamedFstream(master_file_.path.c_str(),
                                                  std::fstream::in);
    return master_file_.r;
  }
  return NULL;
}

Importer::~Importer() {
  master_file_.rw->close();
  master_file_.r->close();
  delete[] & master_file_.path;
}

File: src/importer.rs
pub mod importer {
    use std::fs::File;
    use std::io::prelude::*;
    use std::io::{BufReader};
    use tempfile::tempfile;
    
    struct Import<'a> {
        master_file: &'a File,
        input_file: File,
        //Data about the import
        pub import_level: u32,
        pub import_name: &'a str
    }


    impl<'a> Import <'a> {
        pub fn import(&self, filename:&str, import_level:u32, import_name:&str, import_list:&Vec<Import>) -> &File {
            self.input_file = match File::open(filename) { Ok(file) => file };

            self.master_file = match tempfile() { Ok(file) => &file };
            
            return &self.master_file;            
        }

        fn read(&self) {
            let input_file_bufreader = BufReader::new(self.input_file);
            
            for line in input_file_bufreader.lines() {
                
            }
        }

        fn get_whole_import_command(&self, import_keyword_line:String) {
            
        }

        fn parse_imports(&self, import_command:String) {
            
        }

        fn satisfy_import_scope() {
            
        }

        fn spawn_if_doesnt_exist() {
        }

        fn merge_filestreams() -> String {
            return "ds".to_string();
        }
    }
}
